### PR TITLE
Update bpython to 0.17

### PR DESCRIPTION
**Pull request by flokli**
###### Motivation for this change
As mentioned by @adisbladis, `pythonPackages` is a mess and needs to be cleaned up.

I bumped the `bpython` package to 0.17, cleaned it up, and added `curtsies` and `pyte`, which are both dependencies of `bpython` (at least now)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/pyte/default.nix [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  name  = "${pname}-${version}";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1an54hvyjm8gncx8cgabz9mkpgjkdb0bkyjlkh7g7f94nr3wnfl7";
+  };
+
+  propagatedBuildInputs = [ wcwidth ];
+
+  buildInputs = [ pytest pytestrunner ];
```
Comment (FRidh) [inline pkgs/development/python-modules/pyte/default.nix]: checkInputs

File: pkgs/development/python-modules/bpython/default.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  version = "0.17";
+  name  = "${pname}-${version}";
+
+  # 0.17 is still missing on PyPI, https://github.com/bpython/bpython/issues/706
+  src = fetchurl {
+    url = "https://bpython-interpreter.org/releases/${pname}-${version}.tar.gz";
+    sha256 = "13fyyx06645ikvmj9zmkixr12kzk1c3a3f9s9i2rvaczjycn82lz";
+  };
+
+  buildInputs = [ mock ];
```
Comment (FRidh) [inline pkgs/development/python-modules/bpython/default.nix]: checkInputs

File: pkgs/development/python-modules/bpython/default.nix [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{ stdenv, buildPythonPackage, fetchurl, pygments, greenlet, curtsies, requests, mock }:
+
+buildPythonPackage rec {
+  pname = "bpython";
+  version = "0.17";
+  name  = "${pname}-${version}";
+
+  # 0.17 is still missing on PyPI, https://github.com/bpython/bpython/issues/706
+  src = fetchurl {
```
Comment (FRidh) [inline pkgs/development/python-modules/bpython/default.nix]: fetchPypi

File: pkgs/development/python-modules/curtsies/default.nix [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  name  = "${pname}-${version}";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1vljmw3sy6lrqahhpyg4gk13mzcx3mwhvg8s41698ms3cpgkjipc";
+  };
+
+  propagatedBuildInputs = [ blessings wcwidth pyte ];
+
+  buildInputs = [ nose mock pytest ];
```
Comment (FRidh) [inline pkgs/development/python-modules/curtsies/default.nix]: checkInputs

File: pkgs/development/python-modules/bpython/default.nix [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+  # tests fail: https://github.com/bpython/bpython/issues/712
+  doCheck = false;
+
+  meta = with stdenv.lib; {
+    description = "A fancy curses interface to the Python interactive interpreter";
+    homepage = "https://bpython-interpreter.org/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ flokli ];
+    platforms = platforms.all;
```
Comment (FRidh) [inline pkgs/development/python-modules/bpython/default.nix]: platforms needs to be removed.

File: pkgs/development/python-modules/curtsies/default.nix [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  checkPhase = ''
+    py.test
+  '';
+
+  meta = with stdenv.lib; {
+    description = "Curses-like terminal wrapper, with colored strings!";
+    homepage = https://pypi.python.org/pypi/curtsies;
+    license = licenses.mit;
+    maintainers = with maintainers; [ flokli ];
+    platforms = platforms.all;
```
Comment (FRidh) [inline pkgs/development/python-modules/curtsies/default.nix]: platforms needs to be removed.

File: pkgs/development/python-modules/pyte/default.nix [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  checkPhase = ''
+    py.test -k "not test_input_output"
+  '';
+
+  meta = with stdenv.lib; {
+    description = "Simple VTXXX-compatible linux terminal emulator";
+    homepage = https://github.com/selectel/pyte;
+    license = licenses.lgpl3;
+    maintainers = with maintainers; [ flokli ];
+    platforms = platforms.all;
```
Comment (FRidh) [inline pkgs/development/python-modules/pyte/default.nix]: platforms needs to be removed.

**Review by FRidh — COMMENTED** [review]
platforms is set by `buildPythonPackage`.

File: pkgs/development/python-modules/bpython/default.nix (hunk 3 above)
Comment (flokli) [reply pkgs/development/python-modules/bpython/default.nix]: 0.17 is not released to PyPI yet, thats why `fetchUrl` is used. Issue is linked above.

File: pkgs/development/python-modules/bpython/default.nix (hunk 2 above)
Comment (flokli) [reply pkgs/development/python-modules/bpython/default.nix]: done

File: pkgs/development/python-modules/bpython/default.nix (hunk 5 above)
Comment (flokli) [reply pkgs/development/python-modules/bpython/default.nix]: done

File: pkgs/development/python-modules/curtsies/default.nix (hunk 4 above)
Comment (flokli) [reply pkgs/development/python-modules/curtsies/default.nix]: done

File: pkgs/development/python-modules/curtsies/default.nix (hunk 6 above)
Comment (flokli) [reply pkgs/development/python-modules/curtsies/default.nix]: done

File: pkgs/development/python-modules/pyte/default.nix (hunk 1 above)
Comment (flokli) [reply pkgs/development/python-modules/pyte/default.nix]: done

File: pkgs/development/python-modules/pyte/default.nix (hunk 7 above)
Comment (flokli) [reply pkgs/development/python-modules/pyte/default.nix]: done

**Comment by flokli** [comment]
@FRidh Thanks for your review. I addressed all the problems mentioned.
`bpython-gtk` was removed since `0.14`, and `urwid` is already there, too.
So I added it as `propagatedBuildInput`, and it also runs.

**Comment by FRidh** [comment]
@GrahamcOfBorg python3.pkgs.bpython

**Pull request merged**